Sources/PackLib/Planner.swift

```diff
@@ -12,71 +12,73 @@ public struct Planner: Sendable {
-                case .failure(_ as CancellationError):
-                    // continue loop
-                    break
-                case .failure(let error):
-                    group.cancelAll()
-                    throw error
-                }
-            }
+        let rootPackage = try await RootPackage(
+            from: dependencyData,
+            dumpPackage: self.dumpPackage
+        )
 
-            return packages
-        }
+        let product = try Self.resolveProduct(
+            from: rootPackage,
+            matching: schema.product,
+            type: .application,
+            plist: schema.infoPath,
+            baseBundleID: schema.idSpecifier,
+            iconPath: schema.iconPath,
+            rootResources: schema.resources,
+            entitlementsPath: schema.entitlementsPath
+        )
 
-        var packagesByProductName: [String: String] = [:]
-        for (packageID, package) in packages {
-            for product in package.products ?? [] {
-                packagesByProductName[product.name] = packageID
-            }
+        let extensions = try (schema.extensions ?? []).map {
+            try Self.resolveProduct(
+                from: rootPackage,
+                matching: $0.product,
+                type: .appExtension,
+                plist: $0.infoPath,
+                baseBundleID: $0.bundleID.flatMap(PackSchema.IDSpecifier.bundleID) ?? .orgID(product.bundleID),
+                iconPath: nil,
+                rootResources: $0.resources,
+                entitlementsPath: $0.entitlementsPath
+            )
         }
 
-        let rootPackage = packages[dependencyRoot.identity]!
-        let deploymentTarget = rootPackage.platforms?.first { $0.name == "ios" }?.version ?? "13.0"
-
-        let libraries = rootPackage.products?.filter { $0.type == .autoLibrary } ?? []
+        return Plan(
+            app: product,
+            extensions: extensions
+        )
+    }
 
+    // swiftlint:disable function_parameter_count cyclomatic_complexity
+    private static func resolveProduct(
+        from rootPackage: RootPackage,
+        matching name: String?,
+        type: Plan.ProductType,
+        plist: String?,
+        baseBundleID: PackSchema.IDSpecifier,
+        iconPath: String?,
+        rootResources: [String]?,
+        entitlementsPath: String?
+    ) throws -> Plan.Product {
         let library = try selectLibrary(
-            from: libraries,
-            matching: schema.base.product
+            from: rootPackage.products?.filter { $0.type == .autoLibrary } ?? [],
+            matching: name
         )
-
-        var resources: [Resource] = []
+        var resources: [Plan.Resource] = []
         var visited: Set<String> = []
-        var targets = library.targets.map { (rootPackage, $0) }
+        var targets = library.targets.map { (rootPackage.base, $0) }
         while let (targetPackage, targetName) = targets.popLast() {
```

Comment:
I think the way this is currently written, if an appex and the root app both depend on a `.dynamic`/`.binaryTarget` product, we'll end up packing it twice (same for resources.) We might want to apply some de-duplication logic here, perhaps always packing dependencies in the root app?

Reply:
As long as the current implementation (with duplication) doesn't _break_ anything, I think we can implement deduplication in a followup.